tsdb/shard.go

```diff
@@ -1206,6 +1206,20 @@ func (a Shards) MapType(measurement, field string) influxql.DataType {
 	return typ
 }
 
+func (a Shards) CallType(name string, args []influxql.DataType) (influxql.DataType, error) {
```

Comment:
Why is `CallType` added to the `Shard` type – I do not understand the relevance

Reply:
I'm not sure this is needed anymore.

The original purpose was to be similar to `MapType`. This function was supposed to say, "The shards implement these functions and here are the return types as reported by the storage engine." It still may be worth doing that, but it seems like this method is captured by another struct I created to update the type mapper.

I think I'm going to keep this for consistency and remove the other struct that intercepts this method call.

Reply:
Ok, I finally remember now.

So yes, this is meant to expose the return values for a function and eventually do some type checking too so we can error out earlier in the iterator creation process. The hookups for more user friendly error messages and validation aren't done, but they also aren't that important since later in the system it will find out the types are bad anyway.

The reason this is being added to shard is for consistency. The field mapper that is used in `query/functions.go` wraps a group of shards to expose all of the values from the query engine. Technically, that section of code can handle returning the types since both the storage engine and the query engine are capable of handling call iterators. I like this here though so there's no surprise that the `Shards` type can handle a `mean()` call, but doesn't implement `CallType` to tell the type checker what type it is. I've now refactored it though so they use the same underlying code base though so there is less confusion and less of a chance they will become disconnected.